File: src/sync/status_aggregator.rs
//! 只读汇总持久化事实与运行时同步状态。

use std::sync::atomic::{AtomicU64, Ordering};

use parking_lot::{Mutex, MutexGuard};
use rusqlite::{params, Connection};

use crate::data::repository::{sync_status, transfer_direction};
use crate::error::{AppError, AppResult};
use crate::sync::state::{FailedItem, SyncGlobalState};
use crate::sync::transfer_state::TransferState;

/// SQLite 中没有持久化来源的运行时状态。
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct RuntimeStatus {
    pub editing: u64,
    pub is_running: bool,
    pub last_sync_time: Option<i64>,
    pub is_indexing: bool,
    pub indexing_scanned_folders: u64,
    pub indexing_discovered_items: u64,
    pub content_changed: bool,
    pub sync_phase: Option<String>,
}

impl From<&SyncGlobalState> for RuntimeStatus {
    /// 从完整全局状态中提取运行时字段。
    fn from(state: &SyncGlobalState) -> Self {
        Self {
            editing: state.editing,
            is_running: state.is_running,
            last_sync_time: state.last_sync_time,
            is_indexing: state.is_indexing,
            indexing_scanned_folders: state.indexing_scanned_folders,
            indexing_discovered_items: state.indexing_discovered_items,
            content_changed: state.content_changed,
            sync_phase: state.sync_phase.clone(),
        }
    }
}

impl RuntimeStatus {
    /// 即使数据库快照失败，也要保持生命周期门状态准确。
    pub(crate) fn apply_to(&self, state: &mut SyncGlobalState) {
        state.editing = self.editing;
        state.is_running = self.is_running;
        state.last_sync_time = self.last_sync_time;
        state.is_indexing = self.is_indexing;
        state.indexing_scanned_folders = self.indexing_scanned_folders;
        state.indexing_discovered_items = self.indexing_discovered_items;
        state.content_changed = self.content_changed;
        state.sync_phase.clone_from(&self.sync_phase);
    }
}

/// 在不改写持久化事实的前提下构建完整全局快照。
#[derive(Debug, Default)]
pub struct StatusAggregator {
    next_revision: AtomicU64,
    publication: Mutex<()>,
}

impl StatusAggregator {
    /// 串行化跨 Engine 替换的版本分配与状态发布。
    pub(crate) fn lock_publication(&self) -> MutexGuard<'_, ()> {
        self.publication.lock()
    }

    /// 获取发布锁，并仅在确认竞争时执行回调。
    pub(crate) fn lock_publication_with_contention_hook(
        &self,
        on_contention: impl FnOnce(),
    ) -> MutexGuard<'_, ()> {
        match self.publication.try_lock() {
            Some(guard) => guard,
            None => {
                on_contention();
                self.publication.lock()
            }
        }
    }

    /// 汇总数据库事实与运行时字段，生成新版本快照。
    pub fn snapshot(
        &self,
        conn: &Connection,
        runtime: RuntimeStatus,
    ) -> AppResult<SyncGlobalState> {
        let (total, failed, conflict, uploading, downloading, waiting_network, transfer_failed): (
            i64,
            i64,
            i64,
            i64,
            i64,
            i64,
            i64,
        ) = conn
            .query_row(
                "SELECT
                    (SELECT COUNT(*) FROM sync_items),
                    (SELECT COUNT(*) FROM sync_items WHERE status=?1),
                    (SELECT COUNT(*) FROM sync_items WHERE status=?2),
                    (SELECT COUNT(*) FROM transfer_queue WHERE state=?3 AND direction=?4),
                    (SELECT COUNT(*) FROM transfer_queue
                        WHERE state=?3 AND direction IN (?5, ?6)),
                    (SELECT COUNT(*) FROM transfer_queue WHERE state=?7),
                    (SELECT COUNT(*) FROM transfer_queue WHERE state=?8)",
                params![
                    sync_status::FAILED,
                    sync_status::CONFLICT,
                    i32::from(TransferState::Running),
                    transfer_direction::UPLOAD,
                    transfer_direction::DOWNLOAD,
                    transfer_direction::DOWNLOAD_UPDATE,
                    i32::from(TransferState::WaitingForNetwork),
                    i32::from(TransferState::Failed),
                ],
                |row| {
                    Ok((
                        row.get(0)?,
                        row.get(1)?,
                        row.get(2)?,
                        row.get(3)?,
                        row.get(4)?,
                        row.get(5)?,
                        row.get(6)?,
                    ))
                },
            )
            .map_err(|error| AppError::generic(format!("聚合同步状态失败：{error}")))?;

        let failed_items = {
            let mut statement = conn
                .prepare(
                    "SELECT local_path, error_message
                     FROM sync_items
                     WHERE status=?1
                     ORDER BY local_path
                     LIMIT 20",
                )
                .map_err(|error| AppError::generic(format!("查询失败项失败：{error}")))?;
            let rows = statement
                .query_map(params![sync_status::FAILED], |row| {
                    Ok(FailedItem {
                        relative_path: row.get(0)?,
                        error_message: row.get(1)?,
                    })
                })
                .map_err(|error| AppError::generic(format!("查询失败项失败：{error}")))?;
            let mut items = Vec::new();
            for row in rows {
                items.push(
                    row.map_err(|error| AppError::generic(format!("读取失败项失败：{error}")))?,
                );
            }
            items
        };

        let total = total as u64;
        let failed = failed as u64;
        let conflict = conflict as u64;
        let revision = self.next_revision.fetch_add(1, Ordering::Relaxed) + 1;

        Ok(SyncGlobalState {
            revision,
            total,
            completed: total.saturating_sub(failed + conflict),
            uploading: uploading as u64,
            downloading: downloading as u64,
            waiting_network: waiting_network as u64,
            failed,
            transfer_failed: transfer_failed as u64,
            failed_items,
            conflict,
            editing: runtime.editing,
            is_running: runtime.is_running,
            last_sync_time: runtime.last_sync_time,
            is_indexing: runtime.is_indexing,
            indexing_scanned_folders: runtime.indexing_scanned_folders,
            indexing_discovered_items: runtime.indexing_discovered_items,
            content_changed: runtime.content_changed,
            sync_phase: runtime.sync_phase,
        })
    }
}

File: lib/service/sync/status_aggregator.dart
/// 状态聚合器 —— TaskRunner 快照 + sync_items 计数 → SyncGlobalState。
///
/// 严格对齐 Rust 原版 `src/sync/status_aggregator.rs` + `src/sync/state.rs`：
/// - 一条 SQL 取 7 个计数（total/failed/conflict/uploading/downloading/
///   waiting_network/transfer_failed）
/// - failed_items 按 local_path 字典序 LIMIT 20
/// - completed = total - failed - conflict（saturating）
/// - revision 进程级单调（跨引擎替换仍递增；首快照 revision=1）
/// - publication 互斥锁串行化「版本分配 + 发布」
library;

import 'dart:async';

import 'package:sqflite/sqflite.dart';

import 'package:petal_link/entity/sync_state.dart';
import 'package:petal_link/types/enums.dart';

/// 引擎运行时状态（无 SQLite 持久化来源，只能由引擎运行时注入，
/// 对齐 Rust `RuntimeStatus`）。
class RuntimeStatus {
  /// 被暂停编辑的文件数
  int editing;

  /// 引擎是否正在运行
  bool isRunning;

  /// 上次同步时间（毫秒 epoch）
  int? lastSyncTime;

  /// 是否正在索引云端目录
  bool isIndexing;

  /// 已扫描的文件夹数（索引用）
  int indexingScannedFolders;

  /// 已发现的条目数（索引用）
  int indexingDiscoveredItems;

  /// 是否有目录结构变更
  bool contentChanged;

  /// 当前同步阶段（null = 空闲）
  SyncPhase? syncPhase;

  RuntimeStatus({
    this.editing = 0,
    this.isRunning = false,
    this.lastSyncTime,
    this.isIndexing = false,
    this.indexingScannedFolders = 0,
    this.indexingDiscoveredItems = 0,
    this.contentChanged = false,
    this.syncPhase,
  });

  /// 从完整快照提取运行时字段（对齐 Rust `From<&SyncGlobalState>`）。
  factory RuntimeStatus.from(SyncGlobalState state) {
    return RuntimeStatus(
      editing: state.editing,
      isRunning: state.isRunning,
      lastSyncTime: state.lastSyncTime,
      isIndexing: state.isIndexing,
      indexingScannedFolders: state.indexingScannedFolders,
      indexingDiscoveredItems: state.indexingDiscoveredItems,
      contentChanged: state.contentChanged,
      syncPhase: state.syncPhase,
    );
  }

  /// 拷贝运行时字段到另一实例。
  void copyFrom(RuntimeStatus other) {
    editing = other.editing;
    isRunning = other.isRunning;
    lastSyncTime = other.lastSyncTime;
    isIndexing = other.isIndexing;
    indexingScannedFolders = other.indexingScannedFolders;
    indexingDiscoveredItems = other.indexingDiscoveredItems;
    contentChanged = other.contentChanged;
    syncPhase = other.syncPhase;
  }
}

/// 状态聚合器（对齐 Rust `StatusAggregator`，进程级单例语义）。
///
/// revision 为全局单调计数器：每次 [snapshot] 分配 +1，
/// 与是否广播无关；跨引擎实例共享同一进程级单例时仍保持单调。
class StatusAggregator {
  /// 进程级共享实例（对齐 Rust 全局 STATUS_AGGREGATOR）。
  static final StatusAggregator process = StatusAggregator._();

  StatusAggregator._();

  /// 测试用独立实例（revision 从 0 起）。
  factory StatusAggregator.independent() => StatusAggregator._();

  /// 单调版本计数器
  int _nextRevision = 0;

  /// 发布互斥锁（串行化版本分配 + 发布的 Future 链）
  Future<void> _publication = Future<void>.value();

  /// 在发布屏障内执行 [body]（对齐 Rust `lock_publication`）。
  ///
  /// 同一时刻只有一个发布者能分配 revision 并广播，保证快照全序。
  Future<T> lockPublication<T>(Future<T> Function() body) {
    final completer = Completer<T>();
    _publication = _publication.then((_) async {
      try {
        final result = await body();
        completer.complete(result);
      } catch (e, st) {
        completer.completeError(e, st);
      }
    });
    return completer.future;
  }

  /// 分配下一个单调版本号。
  int nextRevision() => ++_nextRevision;

  /// 聚合 DB 计数与运行时状态，产出权威快照（对齐 Rust `snapshot`）。
  ///
  /// 调用方必须已持有发布锁（经 [lockPublication] 调用）。
  Future<SyncGlobalState> snapshot(Database db, RuntimeStatus runtime) async {
    final rows = await db.rawQuery(
      'SELECT '
      '(SELECT COUNT(*) FROM sync_items) AS total, '
      '(SELECT COUNT(*) FROM sync_items WHERE status = ?) AS failed, '
      '(SELECT COUNT(*) FROM sync_items WHERE status = ?) AS conflict, '
      '(SELECT COUNT(*) FROM transfer_queue WHERE state = ? AND direction = ?) '
      'AS uploading, '
      '(SELECT COUNT(*) FROM transfer_queue WHERE state = ? '
      'AND direction IN (?, ?)) AS downloading, '
      '(SELECT COUNT(*) FROM transfer_queue WHERE state = ?) '
      'AS waiting_network, '
      '(SELECT COUNT(*) FROM transfer_queue WHERE state = ?) '
      'AS transfer_failed',
      [
        SyncItemStatus.failed.code,
        SyncItemStatus.conflict.code,
        TransferState.running.code,
        TransferDirection.upload.code,
        TransferState.running.code,
        TransferDirection.download.code,
        TransferDirection.downloadUpdate.code,
        TransferState.waitingForNetwork.code,
        TransferState.failed.code,
      ],
    );
    final row = rows.first;
    int count(String key) {
      final v = row[key];
      if (v is int) return v;
      return int.tryParse('$v') ?? 0;
    }

    // 失败项详情：按路径字典序排序后截断前 20
    final failedRows = await db.query(
      'sync_items',
      columns: ['local_path', 'error_message'],
      where: 'status = ?',
      whereArgs: [SyncItemStatus.failed.code],
      orderBy: 'local_path ASC',
      limit: SyncGlobalState.maxFailedItems,
    );
    final failedItems = failedRows
        .map((r) => FailedItem(
              relativePath: r['local_path'] as String? ?? '',
              errorMessage: r['error_message'] as String?,
            ))
        .toList();

    final total = count('total');
    final failed = count('failed');
    final conflict = count('conflict');
    final completed = total - failed - conflict;

    return SyncGlobalState(
      revision: nextRevision(),
      total: total,
      completed: completed < 0 ? 0 : completed,
      uploading: count('uploading'),
      downloading: count('downloading'),
      waitingNetwork: count('waiting_network'),
      failed: failed,
      transferFailed: count('transfer_failed'),
      failedItems: failedItems,
      conflict: conflict,
      editing: runtime.editing,
      isRunning: runtime.isRunning,
      lastSyncTime: runtime.lastSyncTime,
      isIndexing: runtime.isIndexing,
      indexingScannedFolders: runtime.indexingScannedFolders,
      indexingDiscoveredItems: runtime.indexingDiscoveredItems,
      contentChanged: runtime.contentChanged,
      syncPhase: runtime.syncPhase,
    );
  }
}
